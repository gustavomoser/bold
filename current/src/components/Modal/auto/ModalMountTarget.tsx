import React, { useEffect, useState } from 'react'

import defaultStore, { ModalStore, ModalStoreItem, ModalStoreState } from './ModalStore'

export interface ModalMountTargetProps {
  store?: ModalStore
}

export function ModalMountTarget(props: ModalMountTargetProps) {
  const { store } = props

  const [items, setItems] = useState<ModalStoreState['items']>([])

  useEffect(() => {
    const unsubstribe = store.subscribe(s => setItems(s.items))
    return () => unsubstribe()
  }, [store])

  return (
    <div>
      {items.map((item: ModalStoreItem) => (
        <React.Fragment key={item.key}>{item.component}</React.Fragment>
      ))}
    </div>
  )
}

ModalMountTarget.defaultProps = {
  store: defaultStore,
} as Partial<ModalMountTargetProps>
