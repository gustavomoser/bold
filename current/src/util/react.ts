import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
