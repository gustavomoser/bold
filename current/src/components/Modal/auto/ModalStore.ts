import { Observable } from '../../../util/Observable'

export interface ModalStoreItem {
  key: string
  component: React.ReactNode
}

export interface ModalStoreState {
  items: ModalStoreItem[]
}

export interface ModalStoreAppendProps {
  dispose(): void
}

export type ModalStoreRenderFn = (props: ModalStoreAppendProps) => React.ReactNode

export class ModalStore extends Observable<ModalStoreState> {
  private value: ModalStoreState = {
    items: [],
  }

  notify(value: ModalStoreState) {
    if (process.env.NODE_ENV === 'development' && this.listeners.length === 0) {
      console.warn(
        `No listeners to the modal store change. ` +
          `Maybe you need to include a <ModalMountTarget /> in your application.`
      )
    }

    super.notify(value)
  }

  append(render: ModalStoreRenderFn) {
    const key = Math.random()
      .toString(36)
      .substring(2)

    const component = render({
      dispose: this.dispose(key),
    })

    this.value = {
      ...this.value,
      items: [...this.value.items, { key, component }],
    }

    this.notify(this.value)

    return key
  }

  private dispose(key: ModalStoreItem['key']) {
    return () => {
      this.value = {
        ...this.value,
        items: this.value.items.filter(item => item.key !== key),
      }
      this.notify(this.value)
    }
  }
}

export default new ModalStore()
