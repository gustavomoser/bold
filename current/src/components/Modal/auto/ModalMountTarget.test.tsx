import { act, render } from '@testing-library/react'
import React from 'react'

import { ModalMountTarget } from './ModalMountTarget'
import { ModalStore } from './ModalStore'

describe('when mounted', () => {
  it('should subscribe to the modal store', () => {
    const store = new ModalStore()
    const subscribeSpy = jest.spyOn(store, 'subscribe')
    render(<ModalMountTarget store={store} />)
    expect(store['listeners']).toHaveLength(1)
    expect(subscribeSpy).toHaveBeenCalled()
  })
})

describe('when unmounted', () => {
  it('should unsubscribe to the modal store', () => {
    const store = new ModalStore()
    const { unmount } = render(<ModalMountTarget store={store} />)
    unmount()
    expect(store['listeners']).toHaveLength(0)
  })
})

describe('render', () => {
  it('should render all components from state', () => {
    const store = new ModalStore()

    const { container } = render(<ModalMountTarget store={store} />)
    expect(container).toMatchSnapshot()

    act(() => store.notify({ items: [{ key: '1', component: <div>First</div> }] }))

    expect(container).toMatchSnapshot()
  })
})
