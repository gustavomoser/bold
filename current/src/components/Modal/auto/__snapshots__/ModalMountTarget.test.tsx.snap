// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render all components from state 1`] = `
<div>
  <div />
</div>
`;

exports[`render should render all components from state 2`] = `
<div>
  <div>
    <div>
      First
    </div>
  </div>
</div>
`;
